crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1179 @@
+    static WINDOWS_PATH_FILTER: [(&str, &str); 1] = [(r"\\\\", "/")];
+
+    #[test]
+    fn constructor_rejects_non_pyi_stdlib_paths() {
+        assert!(ModuleResolutionPathBuf::standard_library("foo.py").is_none());
+        assert!(ModuleResolutionPathBuf::standard_library("foo/__init__.py").is_none());
+        assert!(ModuleResolutionPathBuf::standard_library("foo.py.pyi").is_none());
+    }
+
+    fn stdlib_path_test_case(path: &str) -> ModuleResolutionPathBuf {
+        ModuleResolutionPathBuf::standard_library(path).unwrap()
+    }
+
+    #[test]
+    fn stdlib_path_no_extension() {
+        assert_debug_snapshot!(stdlib_path_test_case("foo"), @r###"
+        ModuleResolutionPath::StandardLibrary(
+            "foo",
+        )
+        "###);
+    }
+
+    #[test]
+    fn stdlib_path_pyi_extension() {
+        assert_debug_snapshot!(stdlib_path_test_case("foo.pyi"), @r###"
+        ModuleResolutionPath::StandardLibrary(
+            "foo.pyi",
+        )
+        "###);
+    }
+
+    #[test]
+    fn stdlib_path_dunder_init() {
+        assert_debug_snapshot!(stdlib_path_test_case("foo/__init__.pyi"), @r###"
+        ModuleResolutionPath::StandardLibrary(
+            "foo/__init__.pyi",
+        )
+        "###);
+    }
+
+    #[test]
+    fn stdlib_paths_can_only_be_pyi() {
+        assert!(stdlib_path_test_case("foo").with_py_extension().is_none());
+    }
+
+    #[test]
+    fn stdlib_path_with_pyi_extension() {
+        assert_debug_snapshot!(
+            ModuleResolutionPathBuf::standard_library("foo").unwrap().with_pyi_extension(),
+            @r###"
+        ModuleResolutionPath::StandardLibrary(
+            "foo.pyi",
+        )
+        "###
+        );
+    }
+
+    #[test]
+    fn non_stdlib_path_with_py_extension() {
+        assert_debug_snapshot!(
+            ModuleResolutionPathBuf::first_party("foo").unwrap().with_py_extension().unwrap(),
+            @r###"
+        ModuleResolutionPath::FirstParty(
+            "foo.py",
+        )
+        "###
+        );
+    }
+
+    #[test]
+    fn non_stdlib_path_with_pyi_extension() {
+        assert_debug_snapshot!(
+            ModuleResolutionPathBuf::first_party("foo").unwrap().with_pyi_extension(),
+            @r###"
+        ModuleResolutionPath::FirstParty(
+            "foo.pyi",
+        )
+        "###
+        );
+    }
+
+    fn non_stdlib_module_name_test_case(path: &str) -> ModuleName {
+        let variants = [
+            ModuleResolutionPathRef::extra,
+            ModuleResolutionPathRef::first_party,
+            ModuleResolutionPathRef::site_packages,
+        ];
+        let results: Vec<Option<ModuleName>> = variants
+            .into_iter()
+            .map(|variant| variant(path).unwrap().to_module_name())
+            .collect();
+        assert!(results
+            .iter()
+            .zip(results.iter().take(1))
+            .all(|(this, next)| this == next));
+        results.into_iter().next().unwrap().unwrap()
+    }
```

Comment:
Uh, I don't think I understand what this method is doing. I think it tests that `to_module_name` of each variant returns the same module name and it then returns one of the module names?

I think writing out the assertions would make debugging this test much easier (and there are only three variants!). I would also take the expected module name as an argument, again, simplifying the code


```suggestion
    fn assert_non_stdlib_module_name(path: &str, expected_name: &ModuleName) {    
        assert_eq!(&ModuleResolutionPathRef::extra(path).unwrap().to_module_name(), expected_module_name);
        assert_eq!(&ModuleResolutionPathRef::first_party(path).unwrap().to_module_name(), expected_module_name);
        assert_eq!(&ModuleResolutionPathRef::site_packages(path).unwrap().to_module_name(), expected_module_name);
    }
```